### PR TITLE
update react-router example to v6

### LINKED ISSUE #8885: Update the React-Router example to the latest version i.e. v5 or v6
Currently the [React Router](https://github.com/carbon-design-system/carbon/tree/main/packages/react/examples/react-router) example is having version: `react-router-dom": "^4.3.1`.

### Summary

- Will update the example as there are some major API changes while migrating to v5 or above compared to v4.
- Will also include Private/Protected Routing in the example itself so the developers can get better understanding overall.
- Would love to add some newer components too in this example, so to keep things up to date.

### Justification

As a developer exploring any open source library, one of the first things one searches for is `example` folder. This PR will help to keep things up-to-date!

### Available extra resources

- Can you affirm what version shall I opt for? (v5 or v6) 
`v6` is going to be released soon and already have started being used by many.

- Can you affirm what must-add components shall I include in the example?

_Carbon is a collaborative system. We encourage teams to build components and
submit them for integration as either add-ons or core components._

Looking forward to be an active contributor to the community. Will be active in raising PR and fixing this.
Thanks,
sohamsshah

### PR DESCRIPTION
**Pull request by sohamsshah**
Closes #8885

#### Changelog

**New**

- Upgrade APIs in React-router example to v6

**Changed**

- Changed APIs

**Removed**

- Removed Older version code and altered as per requirements
 
Will be quick in responding to comments. This is my first contribution to Carbon and super excited to get it merged :)
Thanks!

### CODE REVIEW

File: packages/react/examples/react-router/package.json [hunk 1]
```diff
@@ -7,10 +7,12 @@
     "carbon-components-react": "latest",
     "carbon-icons": "latest",
     "node-sass": "^4.11.0",
-    "react": "^16.7.0",
-    "react-dom": "^16.7.0",
-    "react-router-dom": "^4.3.1",
-    "react-scripts": "2.1.3"
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2",
+    "history": "^5.0.0",
```
Comment (tay1orjones) [inline packages/react/examples/react-router/package.json]: Why is `history` added?

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Thanks for the PR @sohamsshah! 

One question below on dependencies. The `yarn.lock` needs updated as well. I'm also unable to get the example project running.

File: packages/react/examples/react-router/package.json (hunk 1 above)
Comment (sohamsshah) [reply packages/react/examples/react-router/package.json]: Without history, the `react-router-dom` didn't work and gave dependancy error. So added that.

**Comment by sohamsshah** [comment]
Hey, I have added `package.lock` file for it. Should work now, can you please affirm?

**Comment by tw15egan** [comment]
@sohamsshah thanks for hopping in here and helping out! We are using `yarn` throughout the project, so I don't think we want to remove the `yarn.lock` file

**Comment by sohamsshah** [comment]
Okay, i will change it to yarn. Thank you 😋

**Comment by sohamsshah** [comment]
Hi Team, I have updated the `yarn` issue. Pardon for late response. Thank you

**Review by tay1orjones — APPROVED** [review]
Really sorry for the delay here! This got buried in my notifications by accident.

I pulled down the PR, ran `yarn install` in the react-router example directory, then ran `yarn start` and got the following error message:

<details>
  <summary>Click here to expand the super long message</summary>

```
There might be a problem with the project dependency tree.
It is likely not a bug in Create React App, but something you need to fix locally.

The react-scripts package provided by Create React App requires a dependency:

  "babel-jest": "^26.6.0"

Don't try to install it manually: your package manager does it automatically.
However, a different version of babel-jest was detected higher up in the tree:

  /Users/taylorjones/dev/carbon/monorepo/node_modules/babel-jest (version: 27.0.6)

Manually installing incompatible versions is known to cause hard-to-debug issues.

If you would prefer to ignore this check, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
That will permanently disable this message but you might encounter other issues.

To fix the dependency tree, try following the steps below in the exact order:

  1. Delete package-lock.json (not package.json!) and/or yarn.lock in your project folder.
  2. Delete node_modules in your project folder.
  3. Remove "babel-jest" from dependencies and/or devDependencies in the package.json file in your project folder.
  4. Run npm install or yarn, depending on the package manager you use.

In most cases, this should be enough to fix the problem.
If this has not helped, there are a few other things you can try:

  5. If you used npm, install yarn (http://yarnpkg.com/) and repeat the above steps with it instead.
     This may help because npm has known issues with package hoisting which may get resolved in future versions.

  6. Check if /Users/taylorjones/dev/carbon/monorepo/node_modules/babel-jest is outside your project directory.
     For example, you might have accidentally installed something in your home folder.

  7. Try running npm ls babel-jest in your project folder.
     This will tell you which other package (apart from the expected react-scripts) installed babel-jest.

If nothing else helps, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
That would permanently disable this preflight check in case you want to proceed anyway.

P.S. We know this message is long but please read the steps above :-) We hope you find them helpful!

```

</details>

Although If I bypass the preflight check via `SKIP_PREFLIGHT_CHECK=true yarn start` things look to work fine and the app is functional.

**Comment by tay1orjones** [comment]
@sohamsshah we'll need you to sign the DCO as stated in an [earlier comment](https://github.com/carbon-design-system/carbon/pull/8941#issuecomment-863516859) before this will be eligible to merge.

**Comment by sohamsshah** [comment]
> **DCO Assistant Lite bot:** Thanks for your submission! We ask that you all sign our [Developer Certificate of Origin](https://github.com/carbon-design-system/carbon-dco/blob/main/dco.md) before we can accept your contribution. You can sign the DCO by adding a comment below using this text:
> 
> _**I have read the DCO document and I hereby sign the DCO.**_
> 
> **1** out of **2** committers have signed the DCO.
> ✅ @dakahn
> ❌ @sohamsshah
> You can retrigger this bot by commenting **recheck** in this Pull Request

I have read the DCO document and I hereby sign the DCO.

**Comment by sohamsshah** [comment]
> @sohamsshah we'll need you to sign the DCO as stated in an [earlier comment](https://github.com/carbon-design-system/carbon/pull/8941#issuecomment-863516859) before this will be eligible to merge.

Hey I am new on how to sign DCO at carbon...Please let me know if I did it correctly or not

**Comment by tay1orjones** [comment]
recheck

**Comment by tay1orjones** [comment]
@sohamsshah sorry for the issues here with the DCO bot. I think for it to accept it your comment needs to have no other content at all in the comment besides the text "I have read the DCO document and I hereby sign the DCO."

**Comment by sohamsshah** [comment]
I have read the DCO document and I hereby sign the DCO.

**Review by dakahn — APPROVED** [review]

**Comment by sohamsshah** [comment]
Thank you Carbon Team 🤩

**Pull request merged**